src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -0,0 +1,976 @@
+            return "JAVA_FLOAT_UNALIGNED";
+        } else if (type == double.class) {
+            return "JAVA_DOUBLE_UNALIGNED";
+        } else if (type == MemoryAddress.class) {
+            return "ADDRESS_UNALIGNED";
+        } else {
+            throw new IllegalStateException("Unknown type: " + type);
+        }
+    }
+
+    private static Class<?> valueLayoutTypeFor(Class<?> type) {
+        if (type == boolean.class) {
+            return ValueLayout.OfBoolean.class;
+        } else if (type == byte.class) {
+            return ValueLayout.OfByte.class;
+        } else if (type == short.class) {
+            return ValueLayout.OfShort.class;
+        } else if (type == char.class) {
+            return ValueLayout.OfChar.class;
+        } else if (type == int.class) {
+            return ValueLayout.OfInt.class;
+        } else if (type == long.class) {
+            return ValueLayout.OfLong.class;
+        } else if (type == float.class) {
+            return ValueLayout.OfFloat.class;
+        } else if (type == double.class) {
+            return ValueLayout.OfDouble.class;
+        } else if (type == MemoryAddress.class) {
+            return ValueLayout.OfAddress.class;
+        } else {
+            throw new IllegalStateException("Unknown type: " + type);
+        }
+    }
+
+    private void emitInvokeStatic(Class<?> owner, String methodName, String descriptor) {
+        mv.visitMethodInsn(INVOKESTATIC, Type.getInternalName(owner), methodName, descriptor, owner.isInterface());
+    }
+
+    private void emitInvokeInterface(Class<?> owner, String methodName, String descriptor) {
+        mv.visitMethodInsn(INVOKEINTERFACE, Type.getInternalName(owner), methodName, descriptor, true);
+    }
+
+    private void emitInvokeVirtual(Class<?> owner, String methodName, String descriptor) {
+        mv.visitMethodInsn(INVOKEVIRTUAL, Type.getInternalName(owner), methodName, descriptor, false);
+    }
+
+    private void emitGetStatic(Class<?> owner, String fieldName, String descriptor) {
+        mv.visitFieldInsn(GETSTATIC, Type.getInternalName(owner), fieldName, descriptor);
+    }
+
+    private void emitCheckCast(Class<?> cls) {
+        mv.visitTypeInsn(CHECKCAST, Type.getInternalName(cls));
+    }
+
+    private void emitDup(BasicType type) {
+        if (type == BasicType.D || type == BasicType.J) {
+            mv.visitInsn(DUP2);
+        } else {
+            mv.visitInsn(Opcodes.DUP);
+        }
+    }
+
+    /*
+     * Low-level emit helpers.
+     */
+
+    private void emitConstZero(Class<?> type) {
+        emitConst(switch (BasicType.of(type)) {
+            case Z, B, S, C, I -> 0;
+            case J -> 0L;
+            case F -> 0F;
+            case D -> 0D;
+            case L -> null;
+        });
+    }
+
+    private void emitConst(Object con) {
+        if (con == null) {
+            mv.visitInsn(Opcodes.ACONST_NULL);
+            return;
+        }
+        if (con instanceof Integer) {
```

Comment:
those can use the instanceof + type pattern

Reply:
Done